src/PerfTreeParser/PerfTreeParser.class.st

```diff
@@ -0,0 +1,244 @@
+Class {
+	#name : #PerfTreeParser,
+	#superclass : #Object,
+	#instVars : [
+		'content',
+		'totalTime'
+	],
+	#category : #PerfTreeParser
+}
+
+{ #category : #'instance creation' }
+PerfTreeParser class >> fromFile: aFile [
+
+	^ self new
+		  content: aFile asFileReference;
+		  yourself
+]
+
+{ #category : #'instance creation' }
+PerfTreeParser class >> parseFile: aFile [
+
+	^ (self fromFile: aFile) root
+]
+
+{ #category : #accessing }
+PerfTreeParser >> content [
+
+	^ content
+]
+
+{ #category : #accessing }
+PerfTreeParser >> content: aFile [
+	"Extract the interesting content from the file."
+
+	| fileContents i |
+	fileContents := aFile contents lines.
+	i := 1.
+
+	[ (fileContents at: i) beginsWith: '# sample duration' ] whileFalse: [
+		i := i + 1 ].
+
+	self time: (((fileContents at: i) last: 10) first: 7) asNumber.
+
+	content := fileContents reject: [ :line | line beginsWith: '#' ].
+	content := content select: [ :line |
+		           ((line copyWithout: Character space) copyWithout: $|)
+			           isNotEmpty ]
+]
+
+{ #category : #accessing }
+PerfTreeParser >> differenceOfSpacesAt: aNodeIndex [
+	"Gives the difference of the number of spaces between 2 followed lines."
+
+	^ (aNodeIndex between: 1 and: self numberOfFunctions)
+		  ifFalse: [ nil ]
+		  ifTrue: [
+			  (self numberOfSpacesAt: aNodeIndex + 1)
+			  - (self numberOfSpacesAt: aNodeIndex) ]
+]
+
+{ #category : #accessing }
+PerfTreeParser >> findChildrenIndexOfIndex: aNode [
+	"Gives an ordered collection of every indexes of the lines where there is a child."
+
+	| lastNode everyChildren |
+	lastNode := self numberOfFunctions.
+	everyChildren := OrderedCollection new.
+
+	(aNode between: 1 and: lastNode) ifFalse: [ ^ 0 ].
+
+	(((self differenceOfSpacesAt: aNode) between: 1 and: 10) or: [
+		 (self percentageOf: aNode) = (self percentageOf: aNode + 1) ])
+		ifTrue: [
+			everyChildren add: aNode + 1.
+			^ everyChildren ].
+
+	aNode to: lastNode do: [ :i |
+		(self numberOfSpacesAt: i) - (self numberOfSpacesAt: aNode) = 11
+			ifTrue: [ everyChildren add: i ] ].
+
+	^ everyChildren
+]
+
+{ #category : #accessing }
+PerfTreeParser >> findChildrenOfIndex: aNodeIndex [
+	"Gives the nodes representing every children at the node aNodeIndex."
+
+	| collectionOfChildrenIndexes |
+	collectionOfChildrenIndexes := self findChildrenIndexOfIndex: aNodeIndex.
+	^ collectionOfChildrenIndexes collect: [ :childIndex |
+		  self nodeAtIndex: childIndex ]
+]
+
+{ #category : #accessing }
+PerfTreeParser >> lastSpaceAt: aLine [
+	"Gives the index of the last space or '-' at a given line."
+
+	| lastSpace line |
+	(aLine between: 1 and: self numberOfFunctions) ifFalse: [ ^ nil ].
```

Comment:
In many functions you are checking this. Can we filter these undesired cases at the beginning and avoid these weird checks?